src/Hazelcast.Net/Sql/SqlQueryResult.cs

```diff
@@ -0,0 +1,135 @@
+        private readonly Func<Task> _closeFunc;
+
+        private SqlRowMetadata _rowMetadata;
+        private SqlPageEnumerator _pageEnumerator;
+
+        private bool _enumerateStarted;
+        private bool _disposed;
+
+        public SqlRow Current => _pageEnumerator?.Current;
+
+        internal SqlQueryResult(
+            SerializationService serializationService,
+            Task<(SqlRowMetadata rowMetadata, SqlPage page)> fetchFirstTask,
+            Func<Task<SqlPage>> fetchNextFunc,
+            Func<Task> closeFunc)
+        {
+            _serializationService = serializationService;
+            _initTask = fetchFirstTask.ContinueWith(InitFromTaskAsync).Unwrap();
+            _fetchNextFunc = fetchNextFunc;
+            _closeFunc = closeFunc;
+        }
+
+        private async Task InitFromTaskAsync(Task<(SqlRowMetadata rowMetadata, SqlPage page)> fetchFirstTask)
+        {
+            var (rowMetadata, page) = await fetchFirstTask; // Ensure task succeeded or forward exception
+            _rowMetadata = rowMetadata;
+            UpdateCurrentPage(page);
+        }
+
+        private void UpdateCurrentPage(SqlPage page)
+        {
+            _pageEnumerator = new SqlPageEnumerator(_serializationService, _rowMetadata, page);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (!_initTask.IsCompleted)
+                _ = _initTask.ContinueWith(t => t.Exception); // mark possible init exception as observed
+
+            var queryCompleted = _pageEnumerator?.IsLastPage ?? false;
+            if (!_disposed && !queryCompleted)
+                await _closeFunc(); // FIXME [Oleksii] discuss possible exception in Dispose
+
+            _disposed = true;
+        }
+
+        public async ValueTask<bool> MoveNextAsync()
+        {
+            _enumerateStarted = true;
+
+            await _initTask;
+
+            if (_pageEnumerator.MoveNext())
+                return true;
+
+            if (_pageEnumerator.IsLastPage)
+                return false;
+
+            // do not try to continue enumeration if disposed(ing)
+            EnsureNotDisposed();
+
+            var page = await _fetchNextFunc().CfAwait();
+            UpdateCurrentPage(page);
+
+            return _pageEnumerator.MoveNext();
+        }
+
+        public IAsyncEnumerable<SqlRow> EnumerateOnceAsync()
+        {
+            async IAsyncEnumerable<SqlRow> Enumerate()
+            {
+                while (await MoveNextAsync().CfAwait())
+                    yield return Current;
+            }
+
+            EnsureNotDisposed();
+            EnsureEnumerationNotStarted();
+            return Enumerate();
+        }
+
+        public IEnumerable<SqlRow> EnumerateOnce()
+        {
+            IEnumerable<SqlRow> Enumerate()
+            {
+                // FIXME [Oleksii] discuss synchronous approach
+                while (MoveNextAsync().GetAwaiter().GetResult())
```

Comment:
@zpqrtbnk 
We really need synchronous `Enumerate` to support LINQ without external libraries like [System.Linq.Async](https://github.com/dotnet/reactive#systemlinqasync--systeminteractiveasync--systeminteractive), so wanted to check if you're ok with this sync-over-async approach.

Reply:
Well, that's a problem. `GetAwaiter().GetResult()` is bad. Is using System.Linq.Async bad?

Reply:
I don't see anything bad about using it, but anyone who would want to consume SQL result via LINQ will be forced to install separate package. This includes our tests (or in testing library we can just copy/implement needed extension methods).

Reply:
Decided to remove and add example of LINQ over `IAsyncEnumerable` using external package.

Reply:
Done